compiler/rustc_hir_typeck/src/fn_ctxt/_impl.rs

```diff
@@ -138,7 +138,21 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
     #[inline]
     pub fn write_ty(&self, id: hir::HirId, ty: Ty<'tcx>) {
         debug!("write_ty({:?}, {:?}) in fcx {}", id, self.resolve_vars_if_possible(ty), self.tag());
-        self.typeck_results.borrow_mut().node_types_mut().insert(id, ty);
+        let mut typeck = self.typeck_results.borrow_mut();
+        let mut node_ty = typeck.node_types_mut();
+        if let Some(ty) = node_ty.remove(id)
+            && let ty::Error(e) = ty.kind()
```

Comment:
Isn't it enough to check we have an `ErrorGuaranteed` for that?

Reply:
`error_reported()` looks for `ty::Error` even deep within the type, so it would also catch e.g. `(T, {type error})`. I don't know if that's relevant to this check tho